### PR TITLE
feat(data-integrity): Add metric for measuring person data in sync

### PR DESCRIPTION
**Pull request by macobo**
## Problem

For https://github.com/PostHog/posthog/issues/10058
To solve problems we need to know whether our fixes are working.

## Changes

This adds a bit kludgy metrics task which once per day measures 100k people, comparing data in postgres against data in clickhouse and emits statsd metrics for the results. 

You can find a dashboard at https://metrics.posthog.com/d/RZzZ-Mj7z/person-data-integrity?orgId=1

Note that the metric is not intended to be perfect - we can drill into more specific issues going forward.

## How did you test this code?

Running in a production shell.

### CODE REVIEW

File: posthog/tasks/verify_persons_data_in_sync.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        result["total"] += 1
+        pg_person = pg_persons[uuid]
+        if uuid not in ch_persons:
+            result["missing_in_clickhouse"] += 1
+            continue
+        _, ch_version, ch_properties = ch_persons[uuid]
+        ch_properties = json.loads(ch_properties)
+        if ch_version != pg_person.version:
+            result["version_mismatch"] += 1
+            logger.info(
```
Comment (tiina303) [inline posthog/tasks/verify_persons_data_in_sync.py:110]: lets not log this at info level as we know there are lots

File: posthog/tasks/verify_persons_data_in_sync.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                "Found version mismatch",
+                team_id=team_id,
+                uuid=uuid,
+                properties=pg_person.properties,
+                ch_properties=ch_properties,
+            )
+        if pg_person.properties != ch_properties:
```
Comment (tiina303) [inline posthog/tasks/verify_persons_data_in_sync.py:117]: If version mismatches this is pretty much guaranteed to mismatch, so let's ignore this when versions match so we have some useful data from here.

**Review by tiina303 — CHANGES_REQUESTED** [review]
Great stuff here, but I have a few concerns...

Looking at the data https://metrics.posthog.com/d/RZzZ-Mj7z/person-data-integrity?orgId=1 I assume all properties mismatches are due to version mismatch (which has legit reasons - delay). This doesn't help us find problems, when versions match and data doesn't (because we can't distinguish them from the lag), which should never happen and is something we can and should fix.

File: posthog/tasks/verify_persons_data_in_sync.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        ).values_list("id", "uuid", "team_id")[:LIMIT]
+    )
+    person_data.sort(key=lambda row: row[2])  # keep persons from same team together
+
+    results = Counter(
+        {
+            "total": 0,
+            "missing_in_clickhouse": 0,
+            "version_mismatch": 0,
+            "properties_mismatch": 0,
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/tasks/verify_persons_data_in_sync.py:61]: Which ones?

File: posthog/tasks/verify_persons_data_in_sync.py (hunk 2 above)
Comment (macobo) [reply posthog/tasks/verify_persons_data_in_sync.py:117]: Let's not build business logic into these metrics and keep it KISS/as users would observe. Instead if we manage to solve the right problem _and_ they're indeed correlated we will see it from the graphs.

File: posthog/tasks/verify_persons_data_in_sync.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+def verify_persons_data_in_sync() -> None:
+    # :KLUDGE: Rather than filter on created_at directly which is unindexed, we look up the latest value in 'id' column
+    #   and leverage that to narrow down filtering in an index-efficient way
+    max_pk = Person.objects.filter(created_at__lte=now() - PERIOD_START).latest("id").id
+    person_data = list(
+        Person.objects.filter(
+            pk__lte=max_pk, pk__gte=max_pk - LIMIT * 5, created_at__gte=now() - PERIOD_END
+        ).values_list("id", "uuid", "team_id")[:LIMIT]
+    )
+    person_data.sort(key=lambda row: row[2])  # keep persons from same team together
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/tasks/verify_persons_data_in_sync.py:54]: I disagree with your claim that "delay" for syncs is the issue here. Spot-checking the failures, these are still visible in the data right now.

Example: 
- https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgdmVyc2lvbiwgcHJvcGVydGllcyBmcm9tIHBlcnNvbiB3aGVyZSBpZCA9ICcwMTgxNDJjZS0zZGJkLTAwMDAtYzBjYi0xN2VjM2JlYjY3MjMnIiwidGVtcGxhdGUtdGFncyI6e319LCJkYXRhYmFzZSI6NH0sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==
- https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjozLCJuYXRpdmUiOnsicXVlcnkiOiJzZWxlY3QgdmVyc2lvbiwgKiBmcm9tIHBvc3Rob2dfcGVyc29uIHdoZXJlIHV1aWQgPSAnMDE4MTQyY2UtM2RiZC0wMDAwLWMwY2ItMTdlYzNiZWI2NzIzJyIsInRlbXBsYXRlLXRhZ3MiOnt9fSwidHlwZSI6Im5hdGl2ZSJ9LCJkaXNwbGF5IjoidGFibGUiLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=

Yes there are flaws with this method but it's indicating a real problem (as evidenced by the above example). Once we have a solution for the immediate problems we can re-evaluate whether the metric is too noisy.

**Comment by macobo** [comment]
>  I assume all properties mismatches are due to version mismatch (which has legit reasons - delay)

First, why assume? Let's fix the issues and see the metric go down. By baking our assumptions into metrics we just create blind-spots as I've called out multiple times.

Second - looking at the failing persons does show that delay isn't the issue at least for the ones I manually checked.

We should not be over-engineering these metrics and instead be iterating as our needs become more advanced. Note that it's really bad we shipped e.g. https://github.com/PostHog/posthog/pull/10135 before this metric since we have no way of telling whether it improves anything or not. Metrics come first, fixes second.

File: posthog/tasks/verify_persons_data_in_sync.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+from posthog.celery import app
+from posthog.client import sync_execute
+from posthog.models.person import Person
+
+logger = structlog.get_logger(__name__)
+
+# We check up to LIMIT persons between PERIOD_START..PERIOD_END, in batches of BATCH_SIZE
+# This helps keep the metric "moving" as we ship fixes or bugs.
+LIMIT = 100000
```
Comment (yakkomajuri) [inline posthog/tasks/verify_persons_data_in_sync.py:18]: Team Infra has asked us to not fetch more than 10k rows from Postgres at a time - I think that's valid and we should do a paginated fetch here

**Review by yakkomajuri — COMMENTED** [review]
@tiina303 seems deep in the woods here so will leave the bulk of the review to her - happy to provide my opinion if necessary.

Left one comment for now

File: posthog/tasks/verify_persons_data_in_sync.py (hunk 5 above)
Comment (macobo) [reply posthog/tasks/verify_persons_data_in_sync.py:18]: We're fetching only IDs here - tested, this query takes ~50ms and isn't that much data. Loading properties is much heavier.

File: posthog/tasks/verify_persons_data_in_sync.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiina303) [reply posthog/tasks/verify_persons_data_in_sync.py:61]: `created_at` only it seems (we should nuke `is_identified`)

File: posthog/tasks/verify_persons_data_in_sync.py (hunk 2 above)
Comment (tiina303) [reply posthog/tasks/verify_persons_data_in_sync.py:117]: I want to be able to tell if any other fields mismatch, when versions match, if you think this one is important to keep too we can have both, but I need metrics for `data doesn't match for properties or created_at, but versions match` <= as that's something I want to know immediately and find the problem in the code and fix.

**Comment by tiina303** [comment]
> First, why assume? Let's fix the issues and see the metric go down. By baking our assumptions into metrics we just create blind-spots as I've called out multiple times.

I don't understand what blindspots you're talking about, specifically I'm proposing we do:
1. check version - if it doesn't match assume properties nor created_at match (they could match, but we can just assume the worst here) - this is noisy, but we should still keep an eye on it anyway.
2. if version matches - check properties and created_at values - this is a problem we know we need to fix and this isn't noisy as this isn't impacted by delay. I want this to be separately available so we can without noise find these problems fast.

^ where is the blindspot, what are we missing.

**Review by tiina303 — APPROVED** [review]

**Pull request merged**